### DISCUSSION TITLE
Implement e2e-test with real backend API without touching real database, third-party service hitting.

### DISCUSSION THREAD

**Discussion by orlantotes** [cypress-io/cypress#27867] [Questions and Help]
I would like to implement e2e-testing using cypress in our React.js front-end by calling real backend API.
The purpose of this test will be to test backend API through front-end.
But I have no idea how I shoud organize project structure for test.
I know my idea is special case.
I say again, I don't want to change database by calling real API. But to test api in front-end during e2e-test. I want to call real API without mocking api response.
Is it possible in cypress context?
If possible, please help me to structure e2e-test.
Thank you.

**Comment by portyu9** [comment]
Hi @orlantotes 

Yes, but the isolation has to be provided by your **backend/test environment**, not by Cypress itself.

You can absolutely have this flow:

```text
Cypress -> React app -> real backend API
```

while preventing production database/third-party side effects, but the backend needs test-safe dependencies behind it.

A common architecture is:

```text
React app
   |
real backend
   |---- disposable/test database
   |
   `---- fake/sandbox third-party service
```

This still exercises the real frontend/backend HTTP integration. You are only replacing external infrastructure that you do not want the test to mutate.

For database state, use a dedicated E2E database and reset/seed it before tests. Cypress explicitly supports this pattern through `cy.task()`:

```js
beforeEach(() => {
  cy.task('db:seed')
})
```

Alternatively your test environment can expose a test-only seed/reset API and Cypress can call it with `cy.request()`.

For third-party calls made **by your backend**, `cy.intercept()` cannot prevent them because those requests never originate in the browser. Your backend needs to use a sandbox endpoint, fake implementation, WireMock-style server, dependency injection, etc. in the test environment.

So there is no Cypress switch that means "use the real backend but automatically roll back every DB write and third-party call." That boundary has to be designed into the backend/test deployment.

This gives you real end-to-end frontend/backend coverage without touching production data or real external services.

Please mark this answer as accepted if it helped, thank you.
